### PR TITLE
 #216 redirect internal user to event detail when he change participation from link in email invitation

### PR DESCRIPTION
**Pull request by asmaNode**
**Demo**: https://streamable.com/aujioy
**Backend PR** :https://ci.linagora.com/linagora/lgs/openpaas/linagora.esn.calendar/merge_requests/686

**Flow Chart of invite user (internal and external)** 

Flow Chart(invite attendee) (1)

When an orgonizer create an event and invite a list of attendee , an email will be sent with the details of invitation and generate links  for each actions depending of type of the attendee
**External Attendee**
Links action will be generated with excal 
action linksExcal
Each action redirect to EXCAL app (calendar public SPA)

**Internal User**
Links action will be generated with calendar
calendarLink
Each action redirect to Calendar app with new state called participation (with event id)

When the application redirect to participation state  , the modal of detail event will be opened

### CODE REVIEW

**Comment by asmaNode** [comment]
https://ci.linagora.com/linagora/lgs/openpaas/linagora.esn.calendar/merge_requests/686

File: src/linagora.esn.calendar/app/routes.js [hunk 1]
```diff
@@ -32,6 +32,26 @@ function routesConfig($stateProvider) {
...
+      onEnter: function($log, $state, notificationFactory, $location, calOpenEventForm, calEventService) {
+        const calendarHomeId = $location.search().calendarHomeId;
+        const jwt = $location.search().jwt;
+        const eventUid = $location.search().eventUid;
+
+        calEventService.changeParticipationFromLink(jwt).then(response => response)
+          .then(() => calEventService.getEventByUID(calendarHomeId, eventUid));
+        calEventService.getEventByUID(calendarHomeId, eventUid)
+          .then(event => calOpenEventForm(calendarHomeId, event))
+          .catch(err => {
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/routes.js:47]: Do you mean this?

```suggestion
        calEventService.changeParticipationFromLink(jwt).then(response => response)
          .then(() => calEventService.getEventByUID(calendarHomeId, eventUid));
          .then(event => calOpenEventForm(calendarHomeId, event))
          .catch(err => {
```

File: src/esn.calendar.libs/app/services/event-service.js [hunk 2]
```diff
@@ -575,4 +593,7 @@ function calEventService(
       return new CalendarShell(ICAL.Component.fromString(response.data));
     });
   }
+  function getChangeParticipationUrl(jwt) {
+    return `${httpConfigurer.getUrl('/calendar/api/calendars/event/participation')}?jwt=${jwt}`;
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.js:597]: Can we make the `esn-frontend-calendar-public` to use this function also? Right now we have 2 functions that do the exact same thing.

File: src/esn.calendar.libs/app/services/event-service.js [hunk 3]
```diff
@@ -535,6 +539,20 @@ function calEventService(
...
+   * @param {Mixed}    eventUid   the uid of event
+   * @param {Mixed}    jwt        jwt
+   */
+  function changeParticipationFromLink(eventUid, jwt) {
+
+    return $http({ method: 'GET', url: getChangeParticipationUrl(jwt) })
+      .then(() => calendarHomeService.getUserCalendarHomeId())
+      .then(calendarHomeId => getEventByUID(calendarHomeId, eventUid))
+      .then((event, calendarHomeId) => calOpenEventForm(calendarHomeId, event));
+
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.js:553]:
```suggestion
```

File: src/esn.calendar.libs/app/services/event-service.js [hunk 4]
```diff
@@ -535,6 +539,20 @@ function calEventService(
     });
   }
 
+  /**
+   * Change the status of participation from external link
+   * @param {Mixed}    eventUid   the uid of event
+   * @param {Mixed}    jwt        jwt
+   */
+  function changeParticipationFromLink(eventUid, jwt) {
+
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.js:548]:
```suggestion
```

File: src/esn.calendar.libs/app/services/event-service.js [hunk 5]
```diff
@@ -575,4 +593,7 @@ function calEventService(
       return new CalendarShell(ICAL.Component.fromString(response.data));
     });
   }
+  function getChangeParticipationUrl(jwt) {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.js:596]:
```suggestion

  function getChangeParticipationUrl(jwt) {
```

File: src/esn.calendar.libs/app/services/event-service.spec.js [hunk 6]
```diff
@@ -1464,6 +1471,45 @@ describe('The calEventService service', function() {
     // Everything else is covered by the modify fn
   });
 
+  describe('the change participation from link', function() {
+
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.spec.js:1475]:
```suggestion
```

File: src/linagora.esn.calendar/app/routes.js [hunk 7]
```diff
@@ -32,6 +32,25 @@ function routesConfig($stateProvider) {
         }
       }
     })
+    .state('calendar.main.participation', {
+      url: '/participation',
+
```
Comment (huy-ta) [inline src/linagora.esn.calendar/app/routes.js:37]:
```suggestion
```

File: src/esn.calendar.libs/app/services/event-service.spec.js [hunk 8]
```diff
@@ -1464,6 +1471,45 @@ describe('The calEventService service', function() {
...
+              ]
+            ]
+          }]
+        }
+      });
+
+      self.calEventService.changeParticipationFromLink(eventUidd, jwt);
+      self.$httpBackend.flush();
+      expect(calendarHomeServiceMock.getUserCalendarHomeId).to.have.been.called;
+      expect(calOpenEventFormMock).to.have.been.called;
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.spec.js:1509]: We should test it more specifically here. What are the function's parameters?

File: src/esn.calendar.libs/app/services/event-service.spec.js [hunk 9]
```diff
@@ -1464,6 +1471,45 @@ describe('The calEventService service', function() {
     // Everything else is covered by the modify fn
   });
 
+  describe('the change participation from link', function() {
+
+    it('should open modal ', function() {
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.spec.js:1476]:
```suggestion
    it('should open modal after successfully changing the participation status and open the appropriate event dialog', function() {
```

File: src/esn.calendar.libs/app/services/event-service.spec.js [hunk 10]
```diff
@@ -1464,6 +1471,45 @@ describe('The calEventService service', function() {
...
+      const eventUidd = '123';
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.spec.js:1477]:
```suggestion
      const eventUid = '123';
```

File: src/esn.calendar.libs/app/services/event-service.js [hunk 11]
```diff
@@ -535,6 +539,20 @@ function calEventService(
...
   }
 
+  /**
+   * Change the status of participation from external link
+   * @param {Mixed}    eventUid   the uid of event
+   * @param {Mixed}    jwt        jwt
+   */
+  function changeParticipationFromLink(eventUid, jwt) {
+
+    return $http({ method: 'GET', url: getChangeParticipationUrl(jwt) })
```
Comment (huy-ta) [inline src/esn.calendar.libs/app/services/event-service.js:549]: I think we should catch an error here and display a notification, something like "Something went wrong when changing your participation status for the event".

File: src/esn.calendar.libs/app/services/event-service.js (hunk 2 above)
Comment (asmaNode) [reply src/esn.calendar.libs/app/services/event-service.js:597]: yes , we can refactor it

File: src/linagora.esn.calendar/app/routes.js [hunk 12]
```diff
@@ -32,6 +32,24 @@ function routesConfig($stateProvider) {
...
+      resolve: {
+        event: function($log, $state, $location, notificationFactory, calEventService) {
+          const jwt = $location.search().jwt;
+          const eventUid = $location.search().eventUid;
+
+          if (jwt && eventUid) {
+            calEventService.changeParticipationFromLink(eventUid, jwt)
+              .catch(err => {
+                $log.error('Can not display the requested event', err);
+                notificationFactory.weakError(null, 'Can not display the event');
```
Comment (tuanlc) [inline src/linagora.esn.calendar/app/routes.js:46]: Is there translation (i18n) for this?

File: src/linagora.esn.calendar/app/routes.js [hunk 13]
```diff
@@ -32,6 +32,24 @@ function routesConfig($stateProvider) {
         }
       }
     })
+    .state('calendar.main.participation', {
+      url: '/participation',
+      resolve: {
+        event: function($log, $state, $location, notificationFactory, calEventService) {
+          const jwt = $location.search().jwt;
+          const eventUid = $location.search().eventUid;
```
Comment (tuanlc) [inline src/linagora.esn.calendar/app/routes.js:40]:
```suggestion
          const { jwt, eventUid } = $location.search();
```

File: src/linagora.esn.calendar/app/routes.js (hunk 12 above)
Comment (asmaNode) [reply src/linagora.esn.calendar/app/routes.js:46]: yes , i picked it from translation files

**Review by rezk2ll — APPROVED** [review]

**Review by huy-ta — APPROVED** [review]

**Pull request merged**